Add stellar_solana_bridge package

zaelgohary opened the pull request:
### Changes

- Added Bridge package

![image](https://github.com/user-attachments/assets/7627aca4-ffec-4a83-b9ea-1051d133b9c9)


### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3996

### Tested Scenarios

A list of scenarios tried to match the deliverables

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

zaelgohary committed:
Add stellar_solana_bridge package

Commit:
Add missing lang and remove unwanted styles

zaelgohary committed:
Change favicon & index title, add stellar.ts, fix vuetify styles

amiraabouhadid commented:
please fix build errs

amiraabouhadid commented on packages/stellar_solana_bridge/README.md:
```diff
@@ -0,0 +1,5 @@
+# Vue 3 + TypeScript + Vite
```
please update the README file to reflect the bridge's setup and usage instructions

amiraabouhadid commented on packages/stellar_solana_bridge/src/assets/vue.svg:
please remove this

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,64 @@
+const amount = ref(1);
+const transferFee = 50.01;
+
+const validateAmount = (value: number) => {
```
where is checking if user has sufficient balance to cover amount + transfer fees?

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,64 @@
+};
+
+const isValidTransaction = computed(() => {
+  return toAddress.value && amount.value > 0;
```
this only checks if an address is entered but not if it is a valid solana address, can we check if it is a valid solana address too?

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,64 @@
+      <v-text-field v-model="fromAddress" label="Stellar Account Seed" variant="outlined"></v-text-field>
+
+      <!-- To Address -->
+      <v-text-field v-model="toAddress" label="Solana Associated Token Address" variant="outlined"></v-text-field>
```
add rules here to check address validity please if possible , you can also check for address length

amiraabouhadid commented on packages/stellar_solana_bridge/src/services/stellar.ts:
```diff
@@ -0,0 +1,49 @@
+  } catch (error) {
+    console.log(error);
+    console.error("Bridge transfer failed:", (error as unknown as any).response?.data?.extras);
+    throw error;
```
where is this error displayed to the user?

samaradel replied:
it's handled in the UI

amiraabouhadid commented on packages/stellar_solana_bridge/src/services/stellar.ts:
```diff
@@ -0,0 +1,49 @@
+  // Convert Solana address to 32-byte memo
+  const solanaAddressBytes = bs58.decode(solanaRecipientAddress);
+  if (solanaAddressBytes.length !== 32) {
+    throw new Error("Invalid Solana address length");
```
please add this validation in the text-field toaddress rules so user does not have to wait after submitting form to get a validation err

samaradel replied:
it's the decoded one which is not existed in the text filed to validate

AhmedHanafy725 replied:
I prefer that too, you can decode it and make sure the length in bytes is 32

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,64 @@
+      <v-card-title class="text-center">TFT Transfer</v-card-title>
+
+      <!-- From Address -->
+      <v-text-field v-model="fromAddress" label="Stellar Account Seed" variant="outlined"></v-text-field>
```
please add validation rules here too

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,64 @@
+const fromAddress = ref("");
+const toAddress = ref("");
+const amount = ref(1);
+const transferFee = 50.01;
```
please declare this as a const in stellar.ts and use it here

Commit:
Init QR code part (#3998)

* Init QR code part

* - Add confirm dialog after transaction is done
- Update amount validation to take the transfer fees as minimum value

* Update UI

* - Remove unwanted cmd from scripts
- Update the UI with alert
- init the Dockerfile

Commit:
Merge branch 'development' into development_add_solana_bridge

Commit:
Add tft solana docker workflow

AhmedHanafy725 commented on .github/workflows/tft_solana_docker.yml:
```diff
@@ -0,0 +1,36 @@
+name: TFT Solana Docker Build/Push
```
```suggestion
name: Stellar-Solana bridge Docker build/push
```

AhmedHanafy725 commented on packages/stellar_solana_bridge/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM node:18 AS build
+
+WORKDIR /app
+
+COPY . .
+# Install yarn and build project
+RUN yarn install
+RUN yarn lerna run build --no-private && yarn workspace @threefold/tft_solana build
```
this will just build the package and if we run a container from this image it will stop immediately. It should have an entrypoint or CMDs. please check the other packages and how they work. (for example: https://github.com/threefoldtech/tfgrid-sdk-ts/blob/development/packages/playground/Dockerfile)

AhmedHanafy725 commented on packages/stellar_solana_bridge/index.html:
```diff
@@ -0,0 +1,13 @@
+    <meta charset="UTF-8" />
+    <link rel="icon" type="image/svg+xml" href="/favicon.png" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>TFT Solana</title>
```
please update this title
```suggestion
    <title>Stellar-Solana Bridge</title>
```

samaradel replied:
they requested this change in the demo

AhmedHanafy725 commented on packages/stellar_solana_bridge/package.json:
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@threefold/tft_solana",
```
```suggestion
  "name": "@threefold/stellar_solana_bridge",
```

AhmedHanafy725 commented on packages/stellar_solana_bridge/src/assets/vue.svg:
do we use this asset?

samaradel replied:
no, will remove it

AhmedHanafy725 reviewed (changes requested):
when scrolling down, it overflows and can't scroll up again

![image](https://github.com/user-attachments/assets/f764a4d3-f1ea-4937-9005-11f160e809f7)

Commit:
- Remove unwanted icon
- Fix UI
- Validate Solana decoded address in the text field
- Remove style.css file
- Update yarn.lock

Commit:
Update .github/workflows/tft_solana_docker.yml

Co-authored-by: AhmedHanafy725 <41957921+AhmedHanafy725@users.noreply.github.com>

Commit:
Update packages/stellar_solana_bridge/package.json

Co-authored-by: AhmedHanafy725 <41957921+AhmedHanafy725@users.noreply.github.com>

Commit:
- Update Dockerfile
- Update Makefile

AhmedHanafy725 commented on packages/stellar_solana_bridge/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+
+# serve the build files using nginx
+COPY --from=build /app/packages/stellar_solana_bridge/dist /usr/share/nginx/html
+COPY --from=build /app/packages/stellar_solana_bridge/scripts/build-env.sh /usr/share/nginx/html
```
where is that `build-env.sh` file in this package?

Commit:
Update index.html

amiraabouhadid commented:
local build still fails even after pulling most recent changes
<img width="1204" alt="Screenshot 2025-03-26 at 13 04 12" src="https://github.com/user-attachments/assets/20c834c4-2db8-4f47-9799-8cbce8f9d206" />

amiraabouhadid commented:
please update the README file. The make command fails
<img width="962" alt="Screenshot 2025-03-26 at 13 21 58" src="https://github.com/user-attachments/assets/515c7d18-0714-478d-9447-fa4783d395f5" />

Commit:
Fix ui overlapping

Commit:
Update Makefile

AhmedHanafy725 commented on Makefile:
```diff
@@ -14,8 +14,8 @@ ifeq ($(project), playground)
-else ifeq ($(project), UI)
-	cd packages/UI && yarn dev
+else ifeq ($(project), stellar_solana_bridge)
+	cd packages/stellar_solana_bridge && yarn serve
```
I see it's `dev` in package.json
```suggestion
	cd packages/stellar_solana_bridge && yarn dev
```

Commit:
Update Dockerfile and README.md

Commit:
Update Makefile

Co-authored-by: AhmedHanafy725 <41957921+AhmedHanafy725@users.noreply.github.com>

AhmedHanafy725 commented on packages/stellar_solana_bridge/README.md:
```diff
@@ -0,0 +1,31 @@
+# Stellar to Solana Bridge
```
```suggestion
# Stellar-Solana Bridge
```

AhmedHanafy725 commented on packages/stellar_solana_bridge/README.md:
```diff
@@ -0,0 +1,31 @@
+# Stellar to Solana Bridge
+
+This project provides a user-friendly interface for transferring tokens from the Stellar network to the Solana network. The bridge currently supports transfers from Stellar to Solana only.
```
```suggestion
This project provides a user-friendly interface for transferring tokens between the Stellar network and the Solana network. The bridge currently supports transfers from Stellar to Solana only.
```

Commit:
Update README.md

AhmedHanafy725 commented on packages/stellar_solana_bridge/README.md:
````diff
@@ -0,0 +1,30 @@
+1. **Clone the repository**:
+
+```bash
+    git clone https://github.com/threefoldtech/tfgrid-sdk-ts/packages/stellar_solana_bridge
+    cd stellar-solana-bridge
````
this won't work, you should clone the repo and then navigate to the package

Commit:
Update packages/stellar_solana_bridge/README.md

Co-authored-by: AhmedHanafy725 <41957921+AhmedHanafy725@users.noreply.github.com>

Commit:
Update Readme

Commit:
Update packages/stellar_solana_bridge/README.md

Co-authored-by: AhmedHanafy725 <41957921+AhmedHanafy725@users.noreply.github.com>

AhmedHanafy725 commented on packages/stellar_solana_bridge/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY --from=build /app/packages/stellar_solana_bridge/dist /usr/share/nginx/html
+
+WORKDIR /usr/share/nginx/html
+
```
please add bash as it's used in the cmd 

`RUN apk add --no-cache bash`

Commit:
Update Dockerfile

amiraabouhadid commented on packages/stellar_solana_bridge/README.md:
```diff
@@ -0,0 +1,30 @@
+# Stellar-Solana Bridge
+
+This project provides a user-friendly interface for transferring tokens between the Stellar network and the Solana network. The bridge currently supports transfers from Stellar to Solana only.
```
which tokens?

samaradel replied:
TF Tokens

amiraabouhadid commented on packages/stellar_solana_bridge/README.md:
```diff
@@ -0,0 +1,30 @@
+
+## Features
+
+- **QR Code Generation**: Users can scan a QR code with the Threefold Connect app to facilitate the transfer.
```
perhaps put an image of where this QR code is

samaradel replied:
I think it's gonna show the wallet address and not sure about it

amiraabouhadid commented on packages/stellar_solana_bridge/README.md:
````diff
@@ -0,0 +1,30 @@
+
+```bash
+    yarn dev
+```
````
include build commands

amiraabouhadid commented on packages/stellar_solana_bridge/src/components/BridgeSolana.vue:
```diff
@@ -0,0 +1,268 @@
+      class="mb-0"
+      style="max-height: 64px"
+    >
+      This bridge currently only supports Stellar to Solana transfers.
```
```suggestion
      This bridge currently supports only Stellar to Solana transfers.
```

amiraabouhadid reviewed (commented):
<img width="1200" alt="Screenshot 2025-03-26 at 14 49 29" src="https://github.com/user-attachments/assets/207ea837-fcd4-468d-bf4a-698182043f4c" />
playground fails build whenever i run ``yarn lerna run build`` . This happens only on this branch. Please check versions compatibaility and packages

samaradel commented:
@amiraabouhadid please do `yarn && make build` in the root

AhmedHanafy725 committed:
Update lock file

amiraabouhadid commented:
please center the qr code and text below it. Fix the "OR" too please 
<img width="1028" alt="Screenshot 2025-03-26 at 15 13 00" src="https://github.com/user-attachments/assets/2461be4a-65d3-4c68-b96a-4f0313ed90d9" />

Commit:
Update packages/stellar_solana_bridge/src/components/BridgeSolana.vue

Co-authored-by: Amira <56790126+amiraabouhadid@users.noreply.github.com>

Commit:
Update README file

amiraabouhadid commented:
opened up an issue for the respnsiveness #4006
